src/main/java/it/aboutbits/springboot/emailservice/lib/application/EmailParameter.java

```diff
@@ -0,0 +1,62 @@
+package it.aboutbits.springboot.emailservice.lib.application;
+
+import it.aboutbits.springboot.emailservice.lib.AttachmentReference;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotEmpty;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Singular;
+import org.springframework.lang.Nullable;
+
+import java.time.OffsetDateTime;
+import java.util.List;
+import java.util.Set;
+
+@Builder
+public record EmailParameter(
+        @NonNull OffsetDateTime scheduleAt,
+
+        @NonNull Email email,
+
+        @Nullable String reference
+) {
+
+    @Builder
+    public record Email(
+            @NonNull
+            @NotBlank
+            String subject,
+
+            @Singular
+            @NonNull
+            @NotEmpty
+            List<String> recipients,
+
+            @NonNull String textBody,
+            @NonNull String htmlBody,
+
+            @NonNull
+            @NotBlank
+            String fromAddress,
+            @NonNull
+            @NotBlank
+            String fromName,
+
+            @Singular
+            @NonNull
+            Set<Attachment> attachments
+    ) {
+        @Builder
+        public record Attachment(
+                @NonNull
+                AttachmentReference reference,
+                @NonNull
+                @NotBlank
+                String fileName,
+                @NonNull
+                @NotBlank
+                String contentType
+        ) {
+        }
```

Comment:
I would propose that we accept the input stream here and the user only needs to provide in the `AttachmentDataSource` an implementation to save the stream on disk. I think that makes the usage of our service much easier. Like this we can get rid of the reference as well.

Reply:
I don't think so. The email service should not care about file management at all.  
It is the responsibility of the actual application to orchestrate the file storage and optional cleanup.  
All the email service does ist "requesting file payload for sending" and "telling the data source that a file is no longer needed in case it was only temporary"

Reply:
usage would look like this:

```
void sendEmailAction(...) {
  var uniqueFileIdentity = storageService.store(someFile);
  
  emailService.send(email, new Attachment(uniqueFileIdentity));
}
```

```
void releaseAttachment(uniqueFileIdentity) {
  storageService.delete(uniqueFileIdentity);
}

Stream getAttachmentPayload(uniqueFileIdentity) {
  return storageService.getData(uniqueFileIdentity);
}
```

Reply:
I liked the API that we had in EMVA much more. It is a single call and the email service handles it for me. It worked very well in EMVA.

Reply:
Also consider sendOrFail. The email service keeps a clean state. You don't have to do anything from the caller side. Here you would have to clean up manually if it fails.

Reply:
If you sendOrFail, it is your responsibility to handle what happens if it fails.

But consider also this: The mailer service does not know if you are sending a temporary file or a file that should persist after.  
That is why the mailer just reports back that it no longer needs a file but allows the application to implement the necessary logic.

Let me extend my example. The application would implement the interface `AttachmentDataSource`:
```
@Component
@AllArgsConstructor
public class MyAttachmentDataSource implements AttachmentDataSource {
  private final QueryInvoicePdf queryInvoidePdf;

    InputStream getAttachmentPayload(AttachmentReference reference) throws AttachmentException {
      return queryInvoidePdf.payloadByReference(reference);
    }

    void releaseAttachment(AttachmentReference reference) throws AttachmentException {
      // do nothing since we do not want to remove invoices after sending
    }
}
```